MSRIOGroup refactor

dianarg opened the pull request:
- Add more unit tests.
- Use Signal classes to implement MSRIOGroup
- New methods in MSRIO interface to support refactoring.
- Remove old implementation from PlatformIO.
- Closes #435 
- Closes #1040 
- Closes #1042 
- Fixes #997 
- Fixes #821

dianarg committed:
Change MSRIO::add_read to use index, add sample()

- Removes use of raw pointers from the interface.

Change-Id: If68afd51e3138b213f2fe286d8d842e124da748a
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Extend MSRIOGroup unit tests

- Add tests for all signals used in trace and report.
- Change cpuid from KNL to SKX.
- Clean up use of MockPlatformTopo to correctly handle more domains.
- This change should be separate from changes to the MSRIOGroup implementation.

Change-Id: I3618974d36aa4b960ca366de95dfbbbf872e39f4
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Refactor MSRIOGroup to use Signal classes

- Replace use of MSR, MSREncode, and MSRSignal with Signal and its subtypes.
- Add power signals to MSRIOGroup.

Change-Id: I6a58a767b7d2f62d596cd3bb9453340ff1fe3854
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add test helpers for checking aggregation functions

Change-Id: I3d0cf06bb80f2531a6f320d81334c2a36d09f353
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add TimeSignal

- Uses geopm_time() so that all time values are from the same source.
- Implementation similar but not identical to TimeIOGroup.

Change-Id: I68cee08f7ef627e3b7657bc6d4b7fc8451874a3c
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Support power and temperature from MSRIOGroup

- Remove power and temperature combined signals from PlatformIO.
- Rename signals in MSRIOGroup to remove the prefix.

Change-Id: Ia2ad6a1784f54a4eee010a90d5429fb78090e286
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -151,63 +151,131 @@ namespace geopm
+                m_signal_domain[signal_name] = energy_domain;
+                m_signal_units[signal_name] = MSR::M_UNITS_WATTS;
+                m_signal_agg_func[signal_name] = agg_function(msr_name);
+                // @todo: format
```
these todos can be removed; handled elsewhere

cmcantalupo commented on test/geopm_test_helper.cpp:
```diff
@@ -69,3 +70,36 @@ bool is_format_raw64(std::function<std::string(double)> func)
+
+bool is_agg_sum(std::function<double(const std::vector<double> &)> func)
+{
+    return func(example_data) == geopm::Agg::sum(example_data);
```
could be 

   return (func.target == geopm::Agg::sum);

But this seems fine as it

dianarg commented on test/MSRIOGroupTest.cpp:
```diff
@@ -222,55 +235,187 @@ TEST_F(MSRIOGroupTest, supported_cpuid)
+    EXPECT_TRUE(is_agg_sum(func));
+    /// @todo: what should this be?
+    //func = m_msrio_group->agg_function("TIMESTAMP_COUNTER");
+    //EXPECT_TRUE(is_agg_sum(func));
```
see #1058

dianarg commented on test/MSRIOGroupTest.cpp:
```diff
@@ -222,55 +235,187 @@ TEST_F(MSRIOGroupTest, supported_cpuid)
+    for (const auto &name : si_alias) {
+        func = m_msrio_group->format_function(name);
+        EXPECT_TRUE(is_format_double(func));
+
```
extra line

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
     }
 
-    void MSRIOGroup::register_raw_msr_signal(const std::string &msr_name, const MSR &msr_ptr)
+    MSRIOGroup::~MSRIOGroup()
```
could use = default

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
+                std::vector<std::shared_ptr<Signal> > result(num_domain);
+                for (int domain_idx = 0; domain_idx < num_domain; ++domain_idx) {
+                    int max_idx = *m_platform_topo.domain_nested(max_domain, read_domain, domain_idx).begin();
+                    auto min = max_it->second[max_idx];
```
confusing name

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
+            auto read_it = m_signal_available.find(msr_name);
+            if (max_it != m_signal_available.end() &&
+                read_it != m_signal_available.end())
+            {
```
move up

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
+                m_signal_agg_func[signal_name] = agg_function(msr_name);
+            }
+        }
+        m_signal_desc_map["MSR::TEMPERATURE_CORE"] = "Core temperature in degrees C";
```
could move up into loop if temp_signals also contained description

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
+        m_signal_desc_map[time_name] = "Time in seconds used to calculate power";
+
+        // hardcoded parameters for read_signal for power
+        int num_read_loops = 8;
```
derivative_window

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -168,46 +171,116 @@ namespace geopm
+            {"MSR::POWER_PACKAGE", "ENERGY_PACKAGE"},
+            {"MSR::POWER_DRAM", "ENERGY_DRAM"}
+        };
+        m_signal_desc_map["MSR::POWER_PACKAGE"] = "Average package power in watts over the last 80 ms";
```
40 ms not accurate for push_signal for some agents.
"Average package power in watts over a window of 40ms or 8 control loop iterations"

dianarg commented on src/MSRIOGroup.cpp:
```diff
@@ -752,66 +708,32 @@ namespace geopm
-        register_msr_signal(msr_name, msr_name);
-    }
-
     void MSRIOGroup::register_msr_signal(const std::string &signal_name, const std::string &msr_name_field)
```
name that reflects this is for aliases

dianarg commented on src/MSRIOGroup.hpp:
```diff
@@ -127,13 +121,45 @@ namespace geopm
+            // to Signal ptr.  This pointer should be copied when signal
+            // is pushed and used directly for read_signal.
+            std::map<std::string, std::vector<std::shared_ptr<Signal> > > m_signal_available;
+            std::map<std::string, int> m_signal_domain;
```
could be a struct

dianarg commented on src/MSRIOGroup.hpp:
```diff
@@ -144,22 +170,15 @@ namespace geopm
-            std::map<std::string, std::vector<std::shared_ptr<MSRSignal> > > m_name_cpu_signal_map;
             std::map<std::string, std::vector<std::shared_ptr<MSRControl> > > m_name_cpu_control_map;
-            // Pushed signals and controls only
-            std::vector<std::shared_ptr<MSRSignal> > m_active_signal;
```
okay to get rid of MSRSignal class
replaced by RawMSRSignal, MSRFieldSignal

dianarg reviewed (commented):
- Have register_alias skip the alias instead of throwing if underlying MSR signal can't be found
- Can remove DifferenceConbinedSignal and DerivativeCombinedSignal

dianarg committed:
Address review comments

- Use struct instead of parallel vectors for m_signal_available.
- Clean up todos and whitespace.
- Better names.
- Change alias registration so that aliases are silently skipped when
  basis is missing, instead of throwing an error.

Change-Id: I17d849ac070a0161c114accf4cdb2f72fd539b06
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo reviewed (approved):
LGTM
